File: etc/TestEncryptionUtils.java
package etc;

import javax.crypto.Cipher;
import javax.crypto.CipherOutputStream;
import javax.crypto.SecretKeyFactory;
import javax.crypto.spec.IvParameterSpec;
import javax.crypto.spec.PBEKeySpec;
import javax.crypto.spec.SecretKeySpec;
import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.security.NoSuchAlgorithmException;
import java.security.SecureRandom;
import java.util.HexFormat;
import java.util.Random;
import java.util.regex.Pattern;

import static java.nio.charset.StandardCharsets.UTF_8;
import static javax.crypto.Cipher.DECRYPT_MODE;
import static javax.crypto.Cipher.ENCRYPT_MODE;

/**
 * {@link TestEncryptionUtils} implements encryption and decryption helpers which are compatible with JavaScript
 * equivalents in `test-encryption-utils.ts`.
 */
public class TestEncryptionUtils {

    /**
     * PBE-decrypt data which was previously encrypted via {@link #encrypt(String, InputStream, OutputStream)}.
     *
     * @param passphrase the encryption passphrase.
     * @param input the encrypted data.
     * @param output the decryption target.
     */
    public static void decrypt(final String passphrase, final InputStream input,
            final OutputStream output) {
        try {
            final var iv = new byte[16];
            var count = 0;
            while (count < iv.length) {
                count += input.read(iv, count, iv.length - count);
            }
            cipher(passphrase, iv, DECRYPT_MODE, input, output);
        } catch (final IOException e) {
            throw new RuntimeException(e);
        }
    }

    /**
     * PBE-decrypt a string which was previously encrypted via {@link #encryptUTF8Hex(String, String)}.
     *
     * @param passphrase the encryption passphrase.
     * @param inputHex the encrypted string.
     */
    public static String decryptUTF8Hex(final String passphrase, final String inputHex) {
        try {
            final var output = new ByteArrayOutputStream(1024);
            try (output; final var input = new ByteArrayInputStream(
                    hex.parseHex(whitespace.matcher(inputHex).replaceAll("")))) {
                decrypt(passphrase, input, output);
            }
            return output.toString(UTF_8);
        } catch (final IOException e) {
            throw new RuntimeException(e);
        }
    }

    /**
     * PBE-encrypt data with a given passphrase. Writes encrypted data to the provided output stream consisting of a
     * 16-byte randomly generated cipher initialization vector followed by the encrypted data. Use
     * {@link #decrypt(String, InputStream, OutputStream)} to decrypt this data.
     *
     * @param passphrase the encryption passphrase.
     * @param input the input data.
     * @param output the output target.
     */
    public static void encrypt(final String passphrase, final InputStream input,
            final OutputStream output) {
        try {
            final var iv = new byte[16];
            random.nextBytes(iv);
            output.write(iv);
            cipher(passphrase, iv, ENCRYPT_MODE, input, output);
        } catch (final IOException e) {
            throw new RuntimeException(e);
        }
    }

    /**
     * PBE-encrypt an input string with a passphrase via {@link #encrypt(String, InputStream, OutputStream)}. The string
     * is encoded as UTF-8 and the encrypted data is returned as a hex-encoded string.
     *
     * @param passphrase the encryption passphrase.
     * @param inputString the input string.
     */
    public static String encryptUTF8Hex(final String passphrase, final String inputString) {
        try {
            final var output = new ByteArrayOutputStream(1024);
            try (output; final var input = new ByteArrayInputStream(inputString.getBytes(UTF_8))) {
                encrypt(passphrase, input, output);
            }
            return hex.formatHex(output.toByteArray());
        } catch (final IOException e) {
            throw new RuntimeException(e);
        }
    }

    /**
     * Create a {@link Cipher} and encrypt or decrypt from an {@link InputStream} to an {@link OutputStream}.
     *
     * @param passphrase the passphrase.
     * @param iv the initialization vector.
     * @param mode the cipher mode.
     * @param input the input stream.
     * @param output the output stream.
     */
    private static void cipher(final String passphrase, final byte[] iv, final int mode, final InputStream input,
            final OutputStream output) {
        try {
            final var factory = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256");
            final var keySpec = new PBEKeySpec(passphrase.toCharArray(), encryptionSalt,
                    65_536, 256);
            final var key = new SecretKeySpec(factory.generateSecret(keySpec).getEncoded(), "AES");
            final var cipher = Cipher.getInstance("AES/CBC/PKCS5PADDING");
            cipher.init(mode, key, new IvParameterSpec(iv));
            try (final var cipherOutput = new CipherOutputStream(output, cipher)) {
                final var buffer = new byte[1024];
                var count = 0;
                while (-1 != (count = input.read(buffer))) {
                    cipherOutput.write(buffer, 0, count);
                }
            }
        } catch (final Exception e) {
            throw new RuntimeException(e);
        }
    }

    /**
     * Salt used when encrypting data via {@link #encrypt(String, InputStream, OutputStream)} and
     * {@link #decrypt(String, InputStream, OutputStream)} and their downstream convenience methods.
     */
    private static final byte[] encryptionSalt =
            "d*X8RHXD%6tSDcjJcymQcRyUSn7^*BBGCeGvv1CgCmsw8EY&BCsavBAafE4$UqgB".getBytes(UTF_8);

    /**
     * Hex codec.
     */
    private static final HexFormat hex = HexFormat.of();

    /**
     * Random number generator for encryption.
     */
    private static final Random random;

    static {
        try {
            random = SecureRandom.getInstanceStrong();
        } catch (final NoSuchAlgorithmException e) {
            throw new RuntimeException(e);
        }
    }

    /**
     * Pattern used to remove whitespace from hex encoded strings.
     */
    private static final Pattern whitespace = Pattern.compile("\\s+");
}

File: __tests__/test-encryption-utils.ts
import * as crypto from "crypto";

/**
 * PBE-decrypt data which was previously encrypted via {@link encrypt}.
 *
 * @param passphrase the encryption passphrase.
 * @param encrypted the encrypted data.
 */
export function decrypt(passphrase: string, encrypted: Buffer) {
    const key = crypto.pbkdf2Sync(passphrase, encryptionSalt, 65_536, 32, "sha256")
    const decipher = crypto.createDecipheriv("aes-256-cbc", key, encrypted.subarray(0, 16));
    const decrypted = decipher.update(encrypted.subarray(16));
    const final = decipher.final();
    const output = Buffer.alloc(decrypted.length + final.length);
    decrypted.copy(output);
    final.copy(output, decrypted.length);
    return output;
}

/**
 * PBE-decrypt a string which was previously encrypted via {@link encryptUTF8Hex}.
 *
 * @param passphrase the encryption passphrase.
 * @param inputHex the encrypted string.
 */
export function decryptUTF8Hex(passphrase: string, inputHex: string) {
    const input = Buffer.from(inputHex.replace(/\s+/g, ""), "hex");
    return decrypt(passphrase, input).toString("utf8");
}

/**
 * PBE-encrypt data with a given passphrase. Returns a buffer consisting of a 16-byte randomly generated cipher
 * initialization vector followed by the encrypted data. Use {@link decrypt} to decrypt this buffer.
 *
 * @param passphrase the encryption passphrase.
 * @param input the input data.
 */
export function encrypt(passphrase: string, input: Buffer) {
    const iv = crypto.randomBytes(16);
    const key = crypto.pbkdf2Sync(passphrase, encryptionSalt, 65_536, 32, "sha256");
    const cipher = crypto.createCipheriv("aes-256-cbc", key, iv);
    const encrypted = cipher.update(input);
    const final = cipher.final();
    const output = Buffer.alloc(16 + encrypted.length + final.length);
    iv.copy(output);
    encrypted.copy(output, 16);
    final.copy(output, 16 + encrypted.length);
    return output;
}

/**
 * PBE-encrypt an input string with a passphrase via {@link encrypt}. The string is encoded as UTF-8 and the encrypted
 * data is returned as a hex-encoded string.
 *
 * @param passphrase the encryption passphrase.
 * @param input the input string.
 */
export function encryptUTF8Hex(passphrase: string, input: string) {
    return encrypt(passphrase, Buffer.from(input, "utf8")).toString("hex");
}

/**
 * Salt used by {@link decrypt} and {@link encrypt}.
 */
const encryptionSalt = "d*X8RHXD%6tSDcjJcymQcRyUSn7^*BBGCeGvv1CgCmsw8EY&BCsavBAafE4$UqgB";
